src/mpid/ch4/src/ch4r_callbacks.c

```diff
@@ -103,77 +98,18 @@ static int handle_unexp_cmpl(MPIR_Request * rreq)
         match_req = (MPIR_Request *) MPIDIG_REQUEST(rreq, req->rreq.match_req);
 
 #ifndef MPIDI_CH4_DIRECT_NETMOD
-        if (unlikely(match_req && MPIDI_REQUEST_ANYSOURCE_PARTNER(match_req))) {
-            anysource_partner = MPIDI_REQUEST_ANYSOURCE_PARTNER(match_req);
-            if (!MPIR_STATUS_GET_CANCEL_BIT(anysource_partner->status)) {
-                mpi_errno = MPID_Cancel_recv(anysource_partner);
-                if (mpi_errno != MPI_SUCCESS) {
-                    goto fn_fail;
-                }
-                /* What should we do if the anysource partner request is not canceled? */
-                MPIR_Assertp(MPIR_STATUS_GET_CANCEL_BIT(anysource_partner->status));
-            }
-            MPIR_Request_free(MPIDI_REQUEST_ANYSOURCE_PARTNER(match_req));
-            MPIDI_REQUEST_ANYSOURCE_PARTNER(match_req) = NULL;
-            MPIDI_REQUEST_ANYSOURCE_PARTNER(anysource_partner) = NULL;
-        }
-#endif /* MPIDI_CH4_DIRECT_NETMOD */
-
-    } else {
-        /* If this message hasn't been matched yet, look for it in the posted queue. */
-        /* MPIDI_CS_ENTER(); */
-        if (root_comm) {
-#ifdef MPIDI_CH4_DIRECT_NETMOD
-            match_req =
-                MPIDIG_dequeue_posted(MPIDIG_REQUEST(rreq, rank),
-                                      MPIDIG_REQUEST(rreq, tag),
-                                      MPIDIG_REQUEST(rreq, context_id),
-                                      &MPIDIG_COMM(root_comm, posted_list));
-#else /* MPIDI_CH4_DIRECT_NETMOD */
-            int continue_matching = 1;
-            while (continue_matching) {
-                match_req =
-                    MPIDIG_dequeue_posted(MPIDIG_REQUEST(rreq, rank),
-                                          MPIDIG_REQUEST(rreq, tag),
-                                          MPIDIG_REQUEST(rreq, context_id),
-                                          &MPIDIG_COMM(root_comm, posted_list));
-
-                if (match_req && MPIDI_REQUEST_ANYSOURCE_PARTNER(match_req)) {
-                    anysource_partner = MPIDI_REQUEST_ANYSOURCE_PARTNER(match_req);
-
-                    mpi_errno = MPIDI_anysource_matched(anysource_partner,
-                                                        MPIDI_REQUEST(rreq, is_local) ?
-                                                        MPIDI_SHM : MPIDI_NETMOD,
-                                                        &continue_matching);
-
-                    MPIR_ERR_CHECK(mpi_errno);
-
-                    MPIR_Request_free(MPIDI_REQUEST_ANYSOURCE_PARTNER(match_req));
-                    MPIDI_REQUEST_ANYSOURCE_PARTNER(match_req) = NULL;
-                    MPIDI_REQUEST_ANYSOURCE_PARTNER(anysource_partner) = NULL;
-                }
-
-                break;
-            }
+        int is_cancelled;
+        MPIDI_anysrc_try_cancel(match_req, &is_cancelled);
+        /* unexpected message should prevent anysrc partner to be created (typical)
+         * or to be progressed (workq)
```

Comment:
I'm not sure I understand what is meant by the workq scenario here. How is the cancellation different?